### PR TITLE
Add `is_causal` API for `TransformerDecoder`

### PR DESCRIPTION
**Pull request by janEbert**
The same API is implemented for `TransformerEncoder`, where this argument is passed through to the sublayers.

### CODE REVIEW

**Review by mikekgfb — APPROVED** [review]

**Comment by mikekgfb** [comment]
@janEbert should this also cover `class Decoder`?

**Comment by janEbert** [comment]
Sorry, I think my message was unclear. I meant that the API is already implemented for `TransformerEncoder`; this just implements the same thing for `TransformerDecoder`.
Did I understand you correctly?

**Comment by mikekgfb** [comment]
@janEbert can you please submit a PR that updates the definition of *_is_causal in docstrings, based on the change in https://github.com/pytorch/pytorch/pull/97214 which replaces the mutual exclusivity and makes *_is_causal a hint.  (Bad things happen when the hint incorrectly asserts the attention mask is causal, when it is not.  The other direction works fine when is_causal is False but the matrix actually contains a causal mask)

**Comment by janEbert** [comment]
@mikekgfb Thanks for merging already. :)
I'll gladly do both things you mentioned, updating the docstring and implementing the API for `Transformer` as well. Cheers!

**Comment by mikekgfb** [comment]
👍

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
